### PR TITLE
Add database schema and datasources

### PR DESCRIPTION
**Pull request by helloseyedjafari**
### Description
This PR is the very beginning of the entire application database schema.
the entire database needed to work with Quran Soras and Ayas is considered.

The tables are pretty self-explanatory but to add more context:
- `calligraphy`: Takes care of all the languages and language types we are going to have in the project.
-- this table is supposed to store Language code (ISA) and writing symbols name. ie: ar_uthmani for the Arabic language with Uthmani writing symbols.
- `name`: This table is to store any name we are going to have. by any name I mean: sora, books and etc.
- `sora`: We store soras specific properties in this one. a few notes about this table: 
-- `Order`, `calligraphy` are unique together. so to block duplicate names for a single sora number in the same language.
- `aya`: As the name suggests it stores every aya of the Quran (not just sora). please note that the aya content is separated to better handle the difference in `calligraphies`

On top of the schemas, you can find the required `DataSource`s to work with the database.
these are AFAIK for now. So in case we need another function/Datasource we can add quickly.

About the datasources please note that all the input functions are `suspend` and the output functions return a `Flow` type. so you should expect observing database in every get function from datasources. That means if not dropped in higher layers the changes on the database will propogate all the way to the top.

I'm still busy writing the tests for the database. please expect them later.

### CODE REVIEW

File: db/src/commonMain/sqldelight/com/islamversity/db/tables.sq [hunk 1]
```diff
@@ -0,0 +1,101 @@
+import com.islamversity.db.model.EntityUUID;
+import com.islamversity.db.model.Calligraphy;
+
+CREATE TABLE IF NOT EXISTS calligraphy (
```
Comment (meghdadfm) [inline db/src/commonMain/sqldelight/com/islamversity/db/tables.sq:11]: add 'name' property for calligraphy tale

File: db/src/commonMain/sqldelight/com/islamversity/db/tables.sq [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ON calligraphy (code);
+
+CREATE VIEW IF NOT EXISTS no_index_calligraphy
+AS
+SELECT id, code FROM calligraphy;
+
+CREATE TABLE IF NOT EXISTS name (
+    index INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+    id TEXT AS EntityUUID NOT NULL,
+    parentId TEXT AS EntityUUID NOT NULL,
```
Comment (meghdadfm) [inline db/src/commonMain/sqldelight/com/islamversity/db/tables.sq]: rowId is more  standard

Comment (meghdadfm) [inline db/src/commonMain/sqldelight/com/islamversity/db/tables.sq]: need relation from parentId to soraId

**Review by meghdadfm — CHANGES_REQUESTED** [review]

File: db/src/commonMain/kotlin/com/islamversity/db/datasource/AyaLocalDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    ): Flow<List<Aya>>
+
+    fun getAyaWithId(
+        entityId: EntityUUID,
+        calligraphy: Calligraphy,
+        context: CoroutineContext = Dispatchers.Default
+    ): Flow<Aya?>
+}
+
+class AyaLocalDataSourceImpl(
```
Comment (rezaiyan) [inline db/src/commonMain/kotlin/com/islamversity/db/datasource/AyaLocalDataSource.kt:42]: Dagger has not been used!

File: db/src/commonMain/kotlin/com/islamversity/db/datasource/CalligraphyLocalDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.take
+import kotlinx.coroutines.withContext
+import kotlin.coroutines.CoroutineContext
+
+interface CalligraphyLocalDataSource {
+    suspend fun insertCalligraphy(
+        id: EntityUUID,
+        code: com.islamversity.db.model.Calligraphy,
```
Comment (rezaiyan) [inline db/src/commonMain/kotlin/com/islamversity/db/datasource/CalligraphyLocalDataSource.kt]: Use `import com.islamversity.db.model.Calligraphy` instead

File: db/src/commonMain/kotlin/com/islamversity/db/datasource/SoraLocalDataSource.kt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    fun getSoraWithOrderAndCalligraphy(
+        order: Long,
+        calligraphy: Calligraphy,
+        context: CoroutineContext = Dispatchers.Default
+    ): Flow<Sora?>
+}
+
+class SoraLocalDataSourceImpl(
+    private val soraQueries: SoraQueries,
+    private val calligraphyQueries: CalligraphyQueries,
```
Comment (rezaiyan) [inline db/src/commonMain/kotlin/com/islamversity/db/datasource/SoraLocalDataSource.kt]: `calligraphyQueries` seems is useless.

File: db-test/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    packagingOptions {
+        exclude("META-INF/NOTICE")
+        exclude("META-INF/NOTICE.txt")
+        exclude("META-INF/LICENSE")
+        exclude("META-INF/LICENSE.txt")
+        exclude("META-INF/services/javax.annotation.processing.Processor")
+        exclude("META-INF/maven/com.google.guava/guava/pom.properties")
+        exclude("META-INF/maven/com.google.guava/guava/pom.xml")
+        exclude("META-INF/DEPENDENCIES")
+        exclude("META-INF/slider_release.kotlin_module")
```
Comment (rezaiyan) [inline db-test/build.gradle.kts]: Merge them all `kotlin_module`s into one `exclude("*.kotlin_module")`

File: db-test/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.islamversity.db_test">
+
+    /
```
Comment (rezaiyan) [inline db-test/src/main/AndroidManifest.xml]: ???

File: db-test/src/androidTest/java/com/islamversity/db_test/InsertQuranTextTest.kt [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+import kotlinx.coroutines.test.runBlockingTest
+import kotlinx.serialization.json.Json
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import java.io.File
+import kotlin.system.measureTimeMillis
+
+/*
```
Comment (rezaiyan) [inline db-test/src/androidTest/java/com/islamversity/db_test/InsertQuranTextTest.kt:22]: Why has been commented?

File: db/src/commonMain/kotlin/com/islamversity/db/model/Calligraphy.kt [hunk 9]
```diff
@@ -0,0 +1,89 @@
+package com.islamversity.db.model
+
+import com.squareup.sqldelight.ColumnAdapter
+
+inline class LanguageCode(val lang: String) {
```
Comment (rezaiyan) [inline db/src/commonMain/kotlin/com/islamversity/db/model/Calligraphy.kt:5]: It seems the inline feature must be added to the [gradle](https://github.com/islamversity/Reyan/blob/seyed/add-database-schema-and-datasources/db/build.gradle.kts) file.

File: db/src/commonMain/kotlin/com/islamversity/db/model/Calligraphy.kt [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+    val isValid: Boolean
```
Comment (rezaiyan) [inline db/src/commonMain/kotlin/com/islamversity/db/model/Calligraphy.kt:7]: Can be `private`.

File: db/src/commonMain/kotlin/com/islamversity/db/model/EntityUUID.kt [hunk 11]
```diff
@@ -0,0 +1,18 @@
+package com.islamversity.db.model
+
+import com.squareup.sqldelight.ColumnAdapter
+
+data class EntityUUID(
+    val id  : String
+){
+    val isValid : Boolean
```
Comment (rezaiyan) [inline db/src/commonMain/kotlin/com/islamversity/db/model/EntityUUID.kt]: Useless

File: db/src/commonMain/kotlin/com/islamversity/db/model/EntityUUID.kt [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+        get() = id.isNotBlank()
+}
+
+class EntityUUIDAdapter : ColumnAdapter<EntityUUID, String>{
```
Comment (rezaiyan) [inline db/src/commonMain/kotlin/com/islamversity/db/model/EntityUUID.kt]: Useless

File: db/src/commonMain/kotlin/com/islamversity/db/datasource/AyaLocalDataSource.kt [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    private val ayaQueries: AyaQueries,
+    private val calligraphyQueries: CalligraphyQueries,
+    private val ayaContentQueries: AyaContentQueries
+) : AyaLocalDataSource {
+
+    override suspend fun insertAya(
+        aya: AyaWithFullContent,
+        ayaCalligraphy: No_index_calligraphy,
```
Comment (rezaiyan) [inline db/src/commonMain/kotlin/com/islamversity/db/datasource/AyaLocalDataSource.kt]: Useless.

File: db/src/commonMain/sqldelight/com/islamversity/db/tables.sq (hunk 1 above)
Comment (helloseyedjafari) [reply db/src/commonMain/sqldelight/com/islamversity/db/tables.sq:11]: name?
why do we have name in calligraphy?

File: db/src/commonMain/sqldelight/com/islamversity/db/tables.sq (hunk 2 above)
Comment (helloseyedjafari) [reply db/src/commonMain/sqldelight/com/islamversity/db/tables.sq]: In sqlite `rowId` has a definition and refers to our `index` column
please note `rowId` refers to index (long auto generated value) of a row. but `parentId` refers to our generated UUID for that record

Comment (helloseyedjafari) [reply db/src/commonMain/sqldelight/com/islamversity/db/tables.sq]: I'm not sure how we can do that. please send your ideas.
the reason for the uncertainty is,we said: "we can use `name` column for other tables like books.name as well"
not sure how I can create relations for all of them

File: db/src/commonMain/kotlin/com/islamversity/db/datasource/AyaLocalDataSource.kt (hunk 3 above)
Comment (helloseyedjafari) [reply db/src/commonMain/kotlin/com/islamversity/db/datasource/AyaLocalDataSource.kt:42]: can't use it.
remember in kotlin.common package we can not reference any platform specifics. and `@Inject` is Java only.
We will provide all of the classes that are in `common` package using Dagger modules.

File: db-test/src/androidTest/java/com/islamversity/db_test/InsertQuranTextTest.kt (hunk 8 above)
Comment (helloseyedjafari) [reply db-test/src/androidTest/java/com/islamversity/db_test/InsertQuranTextTest.kt:22]: needs to be changed.
due to the DB schema changes

File: db/src/commonMain/kotlin/com/islamversity/db/model/Calligraphy.kt (hunk 10 above)
Comment (helloseyedjafari) [reply db/src/commonMain/kotlin/com/islamversity/db/model/Calligraphy.kt:7]: I need other people to be able to verify

File: db/src/commonMain/kotlin/com/islamversity/db/model/EntityUUID.kt (hunk 11 above)
Comment (helloseyedjafari) [reply db/src/commonMain/kotlin/com/islamversity/db/model/EntityUUID.kt]: needs to be used by creator of the this class

File: db/src/commonMain/kotlin/com/islamversity/db/model/EntityUUID.kt (hunk 12 above)
Comment (helloseyedjafari) [reply db/src/commonMain/kotlin/com/islamversity/db/model/EntityUUID.kt]: will be used when we create the db

File: db-test/build.gradle.kts (hunk 6 above)
Comment (helloseyedjafari) [reply db-test/build.gradle.kts]: lifesaver

**Review by meghdadfm — APPROVED** [review]

**Pull request merged**